core/src/main/java/feast/core/service/StatsService.java

```diff
@@ -0,0 +1,587 @@
+    this.storeRepository = storeRepository;
+    this.specService = specService;
+    this.featureStatisticsRepository = featureStatisticsRepository;
+  }
+
+  /**
+   * Get {@link DatasetFeatureStatistics} for the requested feature set in the provided datasets or
+   * date range for the store provided. The {@link DatasetFeatureStatistics} will contain a list of
+   * {@link FeatureNameStatistics} for each feature requested. Results retrieved will be cached
+   * indefinitely. To force Feast to recompute the statistics, set forceRefresh to true.
+   *
+   * <p>Only one of datasetIds or startDate/endDate should be provided. If both are provided, the
+   * former will be used over the latter.
+   *
+   * <p>If multiple datasetIds or if the date ranges over a few days, statistics will be retrieved
+   * for each single unit (dataset id or day) and results aggregated across that set. As a result of
+   * this, in such a scenario, statistics that cannot be aggregated will be dropped. This includes
+   * all histograms and quantiles, unique values, and top value counts.
+   *
+   * @param request {@link GetFeatureStatisticsRequest} containing feature set name, subset of
+   *     features, dataset ids or date range, and store to retrieve the data from.
+   * @return {@link GetFeatureStatisticsResponse} containing {@link DatasetFeatureStatistics} with
+   *     the feature statistics requested.
+   * @throws IOException
+   */
+  @Transactional
+  public GetFeatureStatisticsResponse getFeatureStatistics(GetFeatureStatisticsRequest request)
+      throws IOException {
+
+    // Get the stats retriever for the store requested
+    StatisticsRetriever statisticsRetriever = getStatisticsRetriever(request.getStore());
+
+    // 1. Retrieve the feature set spec from the db
+    FeatureSetSpec featureSetSpec;
+    try {
+      featureSetSpec = getFeatureSetSpec(request.getFeatureSetId());
+    } catch (IllegalArgumentException | RetrievalException e) {
+      throw new RetrievalException(
+          String.format("Unable to find feature set %s", request.getFeatureSetId()), e);
+    }
+
+    // 2. Filter out the features requested by the user. If none are provided,
+    // use all features in the feature set.
+    List<String> features = request.getFeatureIdsList();
+    if (features.size() == 0) {
+      features =
+          featureSetSpec.getFeaturesList().stream()
+              .map(FeatureSpec::getName)
+              .collect(Collectors.toList());
+    }
+
+    // 3. Retrieve the statistics from the StatsRetriever.
+    List<List<FeatureNameStatistics>> featureNameStatisticsList = new ArrayList<>();
+    if (request.getDatasetIdsCount() == 0) {
+      // If no dataset provided, retrieve by date
+      long timestamp = request.getStartDate().getSeconds();
+      while (timestamp < request.getEndDate().getSeconds()) {
+        List<FeatureNameStatistics> featureNameStatistics =
+            getFeatureNameStatisticsByDate(
```

Comment:
Would this results in multiple calls to the Feast db and multiple calls to BigQuery? If so, can we use the full time range?

Reply:
No, if the full time range is used, the values cannot be cached.

Reply:
Correct me if I am wrong, but wouldn't the only requirement that the query we send to BigQuery returns multiple results? 

Meaning we check our cache (in one query), we ask BQ for the days that we don't have (in one query), we cache them, and then we return the full set.